#!/usr/bin/env python
"""
Find initial mutual exclusive cliques by aligning
input reads against itself.
"""
__author__ = 'etseng@pacificbiosciences.com'

import os
import time
import networkx as nx
import logging
from pbcore.io import FastaReader
from pbcore.util.Process import backticks
from pbtools.pbtranscript.Utils import real_upath
import pbtools.pbtranscript.ice.pClique as pClique
#from pbtools.pbtranscript.ice.IceUtils import blasr_against_ref  # replacing this with dalign_against_ref
from pbtools.pbtranscript.icedalign.IceDalignUtils import DazzIDHandler, DalignerRunner
from pbtools.pbtranscript.icedalign.IceDalignReader import dalign_against_ref

class IceInit(object):
    """Iterative clustering and error correction."""
    def __init__(self, readsFa, qver_get_func,
        ice_opts, sge_opts):

        self.readsFa = readsFa
        self.qver_get_func = qver_get_func
        self.ice_opts = ice_opts
        self.sge_opts = sge_opts
        self.uc = None

        self.uc = self.init_cluster_by_clique(
            readsFa=readsFa,
            qver_get_func=qver_get_func,
            ice_opts=ice_opts, sge_opts=sge_opts)

    # OBSOLETE version using BLASR
    # def _align(self, queryFa, targetFa, outFN, ice_opts, sge_opts):
    #     """Align input reads against itself using BLASR."""
    #     if os.path.exists(outFN):
    #         logging.info("{0} already exists. No need to run BLASR.".
    #             format(outFN))
    #     else:
    #         cmd = "blasr {q} ".format(q=real_upath(queryFa)) + \
    #               "{t} ".format(t=real_upath(targetFa)) + \
    #               "-m 5 -maxLCPLength 15 " + \
    #               "-nproc {cpu} ".format(cpu=sge_opts.blasr_nproc) + \
    #               "-maxScore {score} ".format(score=ice_opts.maxScore) + \
    #               "-bestn {n} -nCandidates {n} ".format(n=ice_opts.bestn) + \
    #               "-out {o}".format(o=real_upath(outFN))
    #         logging.info("Calling {cmd}".format(cmd=cmd))
    #         _output, code, msg = backticks(cmd)
    #         if code != 0:
    #             errMsg = "{cmd} exited with {code}: {msg}".\
    #                     format(cmd=cmd, code=code, msg=msg)
    #             logging.error(errMsg)
    #             raise RuntimeError (errMsg)

    def _align(self, queryFa, output_dir, ice_opts, sge_opts):

        input_obj = DazzIDHandler(queryFa, False)
        DalignerRunner.make_db(input_obj.dazz_filename)

        # run this locally
        runner = DalignerRunner(queryFa, queryFa, is_FL=True, same_strand_only=True, \
                            query_converted=True, db_converted=True, query_made=True, \
                            db_made=True, use_sge=False, cpus=4)
        las_filenames, las_out_filenames = runner.runHPC(min_match_len=800, output_dir=output_dir)
        return input_obj, las_out_filenames


    # OBSOLETE version using BLASR
    # def _makeGraphFromM5(self, m5FN, qver_get_func, ice_opts):
    #     """Construct a graph from a BLASR M5 file."""
    #     alignGraph = nx.Graph()
    #
    #     for r in blasr_against_ref(output_filename=m5FN,
    #         is_FL=True,
    #         sID_starts_with_c=False,
    #         qver_get_func=qver_get_func,
    #         ece_penalty=ice_opts.ece_penalty,
    #         ece_min_len=ice_opts.ece_min_len):
    #         if r.qID == r.cID:
    #             continue # self hit, ignore
    #         if r.ece_arr is not None:
    #             logging.debug("adding edge {0},{1}".format(r.qID, r.cID))
    #             alignGraph.add_edge(r.qID, r.cID)
    #     return alignGraph

    def _makeGraphFromLasOut(self, las_out_filenames, dazz_obj, qver_get_func, ice_opts):
        alignGraph = nx.Graph()

        for las_out_filename in las_out_filenames:
            count = 0
            start_t = time.time()
            for r in dalign_against_ref(dazz_obj, dazz_obj, las_out_filename, is_FL=True, sID_starts_with_c=False,
                      qver_get_func=qver_get_func, qv_prob_threshold=.03,
                      ece_penalty=ice_opts.ece_penalty, ece_min_len=ice_opts.ece_min_len,
                      same_strand_only=True, no_qv_or_aln_checking=False):
                if r.qID == r.cID: continue # self hit, ignore
                if r.ece_arr is not None:
                    alignGraph.add_edge(r.qID, r.cID)
                    count += 1
            logging.debug("total {0} edges added from {1}; took {2} sec".format(count, las_out_filename, time.time()-start_t))
        return alignGraph


    def _findCliques(self, alignGraph, readsFa):
        """
        Find all mutually exclusive cliques within the graph, with decreased
        size.

        alignGraph - a graph, each node represent a read and each edge
        represents an alignment between two end points.

        Return a dictionary of clique indices and nodes.
            key = index of a clique
            value = nodes within a clique
        Cliques are ordered by their size descendingly: index up, size down
        Reads which are not included in any cliques will be added as cliques
        of size 1.
        """
        uc = {}    # To keep cliques found
        used = []  # nodes within any cliques
        ind = 0    # index of clique to discover

        deg = alignGraph.degree().items()
        # Sort tuples of (node, degree) by degree, descendingly
        deg.sort(key=lambda x:x[1], reverse=True)
        for d in deg:
            node = d[0]  # node which has the largest degree in alignGraph
            if node not in alignGraph:
                continue
            # just get the immediate neighbors since we're looking for perfect
            # cliques
            subGraph = alignGraph.subgraph([node] + alignGraph.neighbors(node))
            subNodes = subGraph.nodes()
            # Convert from networkx.Graph to a sparse matrix
            S, H = pClique.convert_graph_connectivity_to_sparse(
                subGraph, subNodes)
            # index of the 'node' in the sub-graph
            seed_i = subNodes.index(node)
            # Grasp a clique from subGraph, and return indices of clique nodes
            tQ = pClique.grasp(S, H, 1., 5, seed_i)
            if len(tQ) > 0:
                c = [subNodes[i] for i in tQ]  # nodes in the clique
                uc[ind] = c  # Add the clique to uc
                ind += 1
                used += c    # Add clique nodes to used
                # Remove clique nodes from alignGraph and continue
                alignGraph.remove_nodes_from(c)

        with FastaReader(readsFa) as reader:
            for r in reader:
                rid = r.name.split()[0]
                if rid not in used:
                    uc[ind] = [rid]
                    ind += 1
        return uc

    def init_cluster_by_clique(self, readsFa, qver_get_func,
            ice_opts, sge_opts):
        """
        Only called once and in the very beginning, when (probably a subset)
        of sequences are given to generate the initial cluster.

        readsFa --- initial fasta filename, probably called *_split00.fa
        qver_get_func --- function that returns QVs on reads
        bestn --- parameter in BLASR, higher helps in finding perfect
            cliques but bigger output
        nproc, maxScore --- parameter in BLASR, set maxScore appropriate
            to input transcript length
        ece_penalty, ece_min_len --- parameter in isoform hit calling

        Self-blasr input then iteratively find all mutually exclusive
            cliques (in decreasing size)
        Returns dict of cluster_index --> list of seqids
        which is the 'uc' dict that can be used by IceIterative
        """
        output_dir = os.path.dirname(readsFa)

        dazz_obj, las_out_filenames = self._align(queryFa=readsFa, output_dir=output_dir,
            ice_opts=ice_opts, sge_opts=sge_opts)

        alignGraph = self._makeGraphFromLasOut(las_out_filenames, dazz_obj, qver_get_func, ice_opts)

        uc = self._findCliques(alignGraph=alignGraph, readsFa=readsFa)
        return uc

